modules/core/src/main/java/org/apache/ignite/internal/util/future/IgniteFutureImpl.java

```diff
@@ -68,7 +80,10 @@ public IgniteFutureImpl(IgniteInternalFuture<V> fut) {
     @Override public void listen(IgniteInClosure<? super IgniteFuture<V>> lsnr) {
         A.notNull(lsnr, "lsnr");
 
-        fut.listen(new InternalFutureListener(lsnr));
+        if (defaultExecutor != null && !isDone())
+            fut.listen(new InternalFutureListener(new AsyncFutureListener<>(lsnr, defaultExecutor)));
+        else
+            fut.listen(new InternalFutureListener(lsnr));
```

Comment:
Also, it is not clean, why `lsnr` used if future `isDone()`

Reply:
Some async operations complete synchronously, like `cache.get` on a primary key. In this case we run the continuation immediately on the current user thread - no need to schedule it somewhere else.

```
// Key 1 is available locally (primary or backup)
// Striped pool is not involved, cache operation and doSomething are executed on the current thread
cache.get(1).listen(doSomething) 
```